File: SilverFactorial64/Sharith/Factorial/Binomial.cs
/// -------- ToujoursEnBeta
/// Author & Copyright : Peter Luschny
/// License: LGPL version 3.0 or (at your option)
/// Creative Commons Attribution-ShareAlike 3.0
/// Comments mail to: peter(at)luschny.de
/// Created: 2010-03-01

#if(MPIR)
using Xint = MpirWrapper.Xint;
#else
using Xint = System.Numerics.BigInteger;
#endif
using Sharith.Math.Primes;

namespace Sharith.Math.Factorial
{
    public sealed class FastBinomial
    {
        private FastBinomial() { }

        public static Xint Binomial(int n, int k)
        {
            if (0 > k || k > n)
            {
                throw new System.ArgumentOutOfRangeException("n",
                 "Binomial: 0 <= k and k <= n required, but n was "
                 + n + " and k was " + k);
            }

            if (k > n / 2) { k = n - k; }

            int rootN = (int)System.Math.Floor(System.Math.Sqrt(n));
            Xint binom = Xint.One;

            var primeCollection = new PrimeSieve(n).GetPrimeCollection();

            foreach (int prime in primeCollection) // Equivalent to a nextPrime() function.
            {
                if (prime > n - k)
                {
                    binom *= prime;
                    continue;
                }

                if (prime > n / 2)
                {
                    continue;
                }

                if (prime > rootN)
                {
                    if (n % prime < k % prime)
                    {
                        binom *= prime;
                    }
                    continue;
                }

                int exp = 0, r = 0, N = n, K = k;

                while (N > 0)
                {
                    r = (N % prime) < (K % prime + r) ? 1 : 0;
                    exp += r;
                    N /= prime;
                    K /= prime;
                }

                if (exp > 0)
                {
                    binom *= Xint.Pow((Xint) prime, exp);
                }
            }
            return binom;
        }

        //static void Main()
        //{
        //    int n = 5673, k = 1239;
        //    // int n = 567, k = 123;

        //    Xint b = Binomial(n, k);
        //    Console.WriteLine("(" + n + "," + k + ") -> " + b);
        //    Console.ReadLine();
        //}
    }
}

File: MpirBasedFunctions/binomial.cpp
// Author & Copyright : Peter Luschny
// Created: 2010-02-01
// License: LGPL version 2.1 or (at your option)
// Creative Commons Attribution-ShareAlike 3.0

// Based on:
// P. Goetgheluck, Computing Binomial Coefficients,
// American Math. Monthly 94 (1987), 360-365.
// with some additional features.

#include <assert.h>
#include "binomial.h"
#include "xmath.h"

void Binomial::PrimeBinomial(Xint binom, ulong n, ulong k)
{
    assert(k <= n);  // TODO: Error handling

    lmp::SetUi(binom, 1);
    if ((k == 0) || (k == n)) return;

    if (k > n / 2) { k = n - k; }
    ulong fi = 0, nk = n - k;

    ulong rootN = Xmath::Sqrt(n);
    ulong* primes;
    ulong piN = Xmath::PrimeSieve(&primes, n);

    for(ulong i = 0; i < piN; i++)
    {
        ulong prime = primes[i];

        if (prime > nk)
        {
            primes[fi++] = prime;
            continue;
        }

        if (prime > n / 2)
        {
            continue;
        }

        if (prime > rootN)
        {
            if (n % prime < k % prime)
            {
                primes[fi++] = prime;
            }
            continue;
        }

        ulong N = n, K = k, p = 1, r = 0;

        while (N > 0)
        {
            r = (N % prime) < (K % prime + r) ? 1 : 0;
            if (r == 1)
            {
                p *= prime;
            }
            N /= prime;
            K /= prime;
        }
        primes[fi++] = p;
    }

    Xmath::Product(binom, primes, 0, fi);
    lmp::FreeUi(primes, piN);
}

